### DISCUSSION TITLE
ref properties: a shortcut for boilerplate read/write properties

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#3528] [Language Ideas]
If I have  a property that wraps some field or other property:

```
public int X
{
	get { return x; }
	set { x = value;}
}
```

I have to write out boilerplate getters and setters. I propose a new syntax "ref properties" that would simplify this, in the same way we have expression bodied properties, but expanded to support setters:

```
public ref int X => x;
```

The `ref` keyword would turn the property from a readonly property into a ref property, which sets x to the value passed in when called as a setter.

Naturally there would need to be some sort of restrictions on the ref values; for instance doing math on them would make little sense.

**Comment by bbarry** [comment]
I think this confusingly overloads the meaning of `ref` used elsewhere. 

While I am not opposed to some nicer syntax for proxy properties, I strongly dislike this particular syntax.

**Comment by jnm2** [comment]
@ekolis The example you have would simplify to `public int X { get; set; }` and I'm not coming up with a more interesting case where an automatically-implemented setter could work. Could you supply some use cases please?

Often the use cases turn out quite nice with #140 which is something that is actively being looked at, but I can't say for sure without seeing one.

Also, the syntax `public ref int X => ref x;` is very similar but it's already legal and it means something else.

**Comment by jnm2** [comment]
Very similar: #1007

**Comment by Joe4evr** [comment]
This has been suggested time and time again, [and the language team has no interest in making mutable properties shorter than immutable ones](https://github.com/dotnet/csharplang/issues/1278#issuecomment-360619835).

**Comment by heku** [comment]
Maybe extend the `ref`'s usage is better:
```C#
public int this[int index] ref _source[index]; 
public int Prop ref _field;
```

**Comment by theunrepentantgeek** [comment]
@heku Spamming the same comment across multiple issues (including [here](https://github.com/dotnet/csharplang/issues/1278#issuecomment-706466030) and [here](https://github.com/dotnet/csharplang/discussions/1007#discussioncomment-94373)] isn't helpful, useful, or welcome.

**Comment by heku** [comment]
> @heku Spamming the same comment across multiple issues (including [here](https://github.com/dotnet/csharplang/issues/1278#issuecomment-706466030) and [here](https://github.com/dotnet/csharplang/discussions/1007#discussioncomment-94373)] isn't helpful, useful, or welcome.

I was not aware of leaving same comment on these similar issues is not helpful and welcome, apologize for that.

**Comment by Rekkonnect** [comment]
I stumbled upon a case where I would actually appreciate its existence, for yet another time.

For the time being, I don't have a better syntax proposal, though I'd avoid usage of `ref`. They would act the same way fields, locals and indexed expressions work; accessibility only matters on the call site of the feature.

Properties that provide a get and set accessor should be available to be passed by "ref". The `init` case is a bit of a corner case, which will be considered.
